### PR TITLE
[JENKINS-26324] Add configurable bridged network mode for Docker containers

### PR DESCRIPTION
**Pull request by ederst**
This pull request adds a new feature described in https://issues.jenkins-ci.org/browse/JENKINS-26324

Changes made:
* ~~Updated Mesos libs to 0.21.0 and HPI plugin to 1.106~~
* HOST and BRIDGE networking for docker containers can now be configured
* Port mappings for those are also possible (if Host Port is left empty, a random Port on the host is used)

~~Snuck in changes (I tried to keep commits atomic, but this happened anyway):~~
* ~~Slave name on Jenkins is now dependent on label (which might not always be good); See: String name = slaveInfo.getLabelString() + "-" + UUID.randomUUID().toString();~~
* ~~The hostname of the docker container is set to the name of the slave~~
* ~~An Exception is now logged if something happens in createMesosTask()~~
* ~~HUDSON_HOME changed from jenkins to /home/jenkins at SLAVE_COMMAND_FORMAT~~

Also: The code for the evaluation of the port range received by the offer is a bit dodgy. I'll try to rewrite this if I got some spare time.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -62,18 +62,15 @@
   </pluginRepositories>
 
   <properties>
-    <!--
-      explicitly specifying the latest version here because one we get from the parent POM
-      tends to lag behind a bit
-    -->
-    <maven-hpi-plugin.version>1.95</maven-hpi-plugin.version>
+    <maven-hpi-plugin.version>1.106</maven-hpi-plugin.version>
```
Comment (KostyaSha) [inline pom.xml]: imho you can update jenkins core version to 1.565.3 that is previous LTS branch version and contains exactly the same hpi-plugin version.

**Comment by KostyaSha** [comment]
Slow logging everywhere.

File: pom.xml (hunk 1 above)
Comment (ederst) [inline pom.xml]: Hmm, yeah. This change was not really meant to be in there as well (sloppy job of me reviewing the changes). I think it is not really necessary for this feature to work. However, I have to test this and clean up my pull request a bit.

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 2]
```diff
@@ -284,6 +308,7 @@ private boolean matches(Offer offer, Request request) {
           "\nRequested for Jenkins slave:\n" +
           "  cpus: " + requestedCpus + "\n" +
           "  mem:  " + requestedMem + "\n" +
+          "  ports: " + hasPortMappings + "\n" +
```
Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: Might be more helpful to know what ports are being requested.

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly [hunk 3]
```diff
@@ -95,6 +95,37 @@
                                 </f:radioBlock>
                             </f:entry>
 
+                            <f:entry title="${%Networking}" description="${%Specify the networking mode to use for this container}" field="networking" >
+                                <f:radioBlock id="networking.HOST" name="networking" title="${%Host}" value="HOST" inline="true" checked="${slaveInfo.containerInfo.networking == 'HOST'}" />
```
Comment (barkerd427) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly]: What about the `container` option?

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java [hunk 4]
```diff
@@ -180,13 +184,27 @@ public String getImage() {
...
       this.dockerImage = dockerImage;
       this.volumes = volumes;
+
+      if (networking == null) {
+          this.networking = Network.BRIDGE.toString();
+      } else {
+          this.networking = networking;
+      }
+
+      if (Network.HOST.equals(Network.valueOf(networking))) {
```
Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: Should a PortMapping perhaps set the host port if the networking is HOST as the mapping is really just implied in Docker?

File: src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly (hunk 3 above)
Comment (barkerd427) [inline src/main/resources/org/jenkinsci/plugins/mesos/MesosCloud/config.jelly]: Disregard. I hadn't realized it wasn't included in Mesos.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java [hunk 5]
```diff
@@ -10,8 +11,9 @@
 import net.sf.json.JSONSerializer;
 
 import org.apache.commons.lang.StringUtils;
-import org.apache.mesos.Protos.ContainerInfo;
+import org.apache.mesos.Protos.ContainerInfo.DockerInfo.Network;
 import org.kohsuke.stapler.DataBoundConstructor;
+import org.springframework.transaction.annotation.Transactional;
```
Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: Why was this added?

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 6]
```diff
@@ -391,11 +444,63 @@ private void createMesosTask(Offer offer, Request request) {
...
+              List<Integer> portsToUse = findPortsToUse(offer, request, portMappings.size());
+
+              Value.Ranges.Builder portRangesBuilder = Value.Ranges.newBuilder();
+
+              for (MesosSlaveInfo.PortMapping portMapping : portMappings) {
+                  PortMapping.Builder portMappingBuilder = PortMapping.newBuilder() //
+                          .setContainerPort(portMapping.getContainerPort()) //
+                          .setProtocol(portMapping.getProtocol());
+
+                  int portToUse = portMapping.getHostPort() == null?portsToUse.get(portToUseIndex++):portMapping.getHostPort();
```
Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: Use spaces around the operators.

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java [hunk 7]
```diff
@@ -71,7 +75,7 @@
   private static final double JVM_MEM_OVERHEAD_FACTOR = 0.1;
 
   private static final String SLAVE_COMMAND_FORMAT =
-      "java -DHUDSON_HOME=jenkins -server -Xmx%dm %s -jar ${MESOS_SANDBOX-.}/slave.jar %s -jnlpUrl %s";
+      "java -DHUDSON_HOME=/home/jenkins -server -Xmx%dm %s -jar ${MESOS_SANDBOX-.}/slave.jar %s -jnlpUrl %s";
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: Why /home/?

**Comment by vinodkone** [comment]
Have all the comments raised so far been addressed? Also, can you please rebase?

**Comment by ederst** [comment]
@vinodkone Hi, will address the comments (and clean the pull request) when I have some spare time. Right now I'm quite busy.

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java (hunk 5 above)
Comment (ederst) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: Oversight, does not need to be there...

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java (hunk 4 above)
Comment (ederst) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: @barkerd427 I'm not sure if I understand you correctly, but if I do then you suggest, adding a port mapping configuration for the HOST network mode, just for the host port?! I don't know if this is allowed by Docker (will have to try this).

File: src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java (hunk 2 above)
Comment (ederst) [inline src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java]: @barkerd427 True that. Realized this when I started a slave with the VNC port exposed and I had no clue to which host port I have to connect. In the meantime I've implemented a feature which shows the (host) ports exposed on the Mesos Slaves on the Jenkins Slave Status page. Still, I have to test this and prepare another pull request...

File: src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java (hunk 4 above)
Comment (ederst) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: @barkerd427 AFAIK now is that HOST mode in Docker exposes the ports of all services in a container on the host network interface. This means that it will not be possible to map ports differently (the ports would have to be configured in the service configuration, for example "Listen 80" in httpd.conf)

Comment (barkerd427) [inline src/main/java/org/jenkinsci/plugins/mesos/MesosSlaveInfo.java]: You're right. I was confused it seems.

**Comment by ederst** [comment]
I'd say all issues are addressed so far, so the pull request can be merged. If not, please tell me why and I'll try to fix it.

**Comment by maselvaraj** [comment]
@ederst Can you please squash the commits ?

**Comment by ederst** [comment]
@maselvaraj Squashed
